### PR TITLE
Align ExpDecay implementation with documentation

### PR DESCRIPTION
**Pull request by DrChainsaw**
Fix for #1176

### CODE REVIEW

File: test/optimise.jl [hunk 1]
```diff
@@ -57,6 +57,17 @@ end
 end
 
 @testset "ExpDecay" begin
+
+	@testset "Sanity Check" begin
+		o = ExpDecay(0.2, 0.5, 1, 1e-3)
```
Comment (dhairyagandhi96) [inline test/optimise.jl:62]: Could we align the formatting with the rest of the file? Good to go otherwise.

Comment (DrChainsaw) [reply test/optimise.jl]: Better now? 

That whole testset has a different intendation compared to the others and I didn't want to obscure the actual changes made.

**Comment by dhairyagandhi96** [comment]
Thoughts on keeping the current set if someone wants the closed form lr decay?

**Comment by DrChainsaw** [comment]
Current set?

I was a bit tempted to rewrite so that the current decayed learning rate was not a state and instead compute it for each individual parameter. Is that what you mean? 

This would be an additional change in behaviour as the current implementation will use the decayed learning rate of the "first seen" parameter for all parameters. 

I guess there are ways this approach can cause unexpected behaviour in general cases and that might be a motivation to change the behaviour.

**Comment by dhairyagandhi96** [comment]
But the lr at any step should be the same for all parameters from the classic definition of ExpDecay. That's how it's implemented in other frameworks as well.

**Comment by DrChainsaw** [comment]
> But the lr at any step should be the same for all parameters from the classic definition of ExpDecay. That's how it's implemented in other frameworks as well.

I guess both the current implementation and the one I proposed above fulfills this requirement in the simple case when the optimizer sees the same set of parameters in each pass. 

In case a new parameter suddenly appears the current implementation will give it the same decayed learning rate as all the other parameters. However, I think that this new parameter will disrupt the learning rate schedule as it will have its own stepcount and eventually trigger a decay based on it which is pretty much out of sync with the other parameters (unless of course the new parameter happens to appear right at a decay step so it is accidentally in sync with the rest). This has nothing to do with the change in this PR as it is a consequence of the mechanism to detect steps which I haven't touched.

The alternative I proposed would just let the new parameter follow its own decay path independently of the existing ones, meaning it will not disrupt the decaying process. 

Again, for the normal case when the set of parameters to update never changes there is no difference.

If we care alot about new parameters entering with the same decayed learning rate as old then I think it will be a bit hard to achive without changing the optimizer API, but maybe there is some clever way to do it.

About that "keeping the current set": I guess this is not what you meant then?

**Comment by dhairyagandhi96** [comment]
I think for now, let's merge this and discuss further cleaning up as a separate PR

bors r+

**Pull request merged**